### PR TITLE
Unify mesh reading dem cfd

### PR DESCRIPTION
**Pull request by blaisb**
### Description

The CFD and the DEM part of the software (lethe-fluid vs lethe-particles) did not follow the same workflow for reading the mesh. Essentially this meant that a lot of features were duplicated between the CFD and the DEM for mesh parsing, but also that a lot of features could not be used for DEM mesh parsing (box refinement, refining at the wall of a domain, etc.). This PR unifies the interface under a unique umbrella. Now all meshes are parsed using the same functionnalities and can use the same feature.

### Testing

No test results should change
A new unit test was added to test the mesh parsing independently.

### Documentation

The fact that the DEM can use the same feature as the CFD for mesh parsing was integrated within the documentation and the release notes.

### Use of LLMs  (Large Language Models)

I did the majority of the work myself but using Claude Code Opus 5.0 to pre-review the PR and update the documentation. 

### Checklist (will be removed when merged)
See [this page](https://chaos-polymtl.github.io/lethe/documentation/contributing/pull-requests.html) for more information about the pull request process.

Code related list:
- [x] All in-code documentation related to this PR is up to date (Doxygen format)
- [x] Copyright headers are present and up to date
- [x] Lethe documentation is up to date
- [x] The branch is rebased onto master
- [x] Code is indented with indent-all and .prm files (examples and tests) with prm-indent
- [x] If parameters are modified, the tests and the documentation of examples are up to date
- [x] An entry describing the change has been added in `/release_notes/current/` following the instructions of `/release_notes/template.md`

Pull request related list:
- [x] No other PR is open related to this refactoring
- [x] Labels are applied
- [x] There are at least 2 reviewers (or 1 if small feature)
- [x] If any future work is planned, an issue is opened
- [x] The PR description is clean and is ready to be used as the commit message when merging the PR

### CODE REVIEW

File: include/core/manifolds.h [hunk 1]
```diff
@@ -52,7 +52,7 @@ namespace Parameters
     // boundary ID. Hence, the total number of individual manifolds should be
     // accessed through the size of the manifold description vectors (e.g.,
     // `ids`).
-    unsigned int number_of_manifolds_subsections;
+    unsigned int number_of_manifolds_subsections = 0;
```
Comment (blaisb) [inline include/core/manifolds.h:55]: I understand that right now we generally don't attribute default value to parameters. But this greatly simplified the implementation and it goes along the way of what we wanted to do (and that I will do) in a follow-up PR which is to encode the default values directly into the struct when possible to enable better unit testing.

File: include/core/parameters.h [hunk 2]
```diff
@@ -29,18 +29,30 @@ namespace Parameters
 {
   struct SizeOfSubsections
   {
-    int boundary_conditions;
-    int manifolds;
+    int boundary_conditions = 0;
+    int manifolds           = 0;
```
Comment (blaisb) [inline include/core/parameters.h:33]: I understand that right now we generally don't attribute default value to parameters. But this greatly simplified the implementation and it goes along the way of what we wanted to do (and that I will do) in a follow-up PR which is to encode the default values directly into the struct when possible to enable better unit testing.

**Comment by blaisb** [comment]
https://github.com/user-attachments/assets/1ea2a5d0-513b-4422-ae3b-73968b1c2328

This also enables DEM simulations with non-uniform mesh btw.

File: include/core/periodic_boundary.h
Comment (karlwilson) [inline include/core/periodic_boundary.h:1]: This is a clean addition, I like it

File: include/core/lethe_grid_tools.h [hunk 3]
```diff
@@ -142,6 +142,26 @@ namespace LetheGridTools
     Point<dim>                                           &point_1,
     Point<dim>                                           &point_2);
 
+  /**
+   * @brief Flag for refinement all the cells of a triangulation which lie within
```
Comment (karlwilson) [inline include/core/lethe_grid_tools.h]:
```suggestion
   * @brief Flag for refinement of all the cells of a triangulation which lie within
```

File: include/core/parameters.h [hunk 4]
```diff
@@ -29,18 +29,32 @@ namespace Parameters
...
   };
 
 
   /**
    * @brief Extract the maximum number of all variable size sections within the parameter file
    *
-   * @param file_name Name of the parameter file from which the size are parsed
+   * @param[in] file_name Name of the parameter file from which the size are
```
Comment (karlwilson) [inline include/core/parameters.h]:
```suggestion
   * @param[in] file_name Name of the parameter file from which the sizes are
```

File: applications/lethe-parameter-check/lethe_parameter_check.cc [hunk 5]
```diff
@@ -54,18 +54,24 @@ main(int argc, char *argv[])
         }
       else if (solver_family == "lethe-particles")
         {
+          // The DEM solvers do not require any variable size subsection. Their
+          // boundary conditions are declared in the "DEM boundary conditions"
+          // subsection, and a file that declares no manifold leads to
+          // subsections of size zero.
+          const Parameters::SizeOfSubsections size_of_subsections =
```
Comment (antonellasuccar) [inline applications/lethe-parameter-check/lethe_parameter_check.cc]: DEM solvers or DEM solver? Technically, it's one solver, right?
```suggestion
          // The DEM solver does not require any variable size subsection. Its
          // boundary conditions are declared in the "DEM boundary conditions"
          // subsection, and a file that declares no manifold leads to
          // subsections of size zero.
```

File: doc/source/parameters/dem/mesh.rst [hunk 6]
```diff
@@ -1,8 +1,19 @@
+..
+  SPDX-FileCopyrightText: Copyright (c) 2026 The Lethe Authors
+  SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception OR LGPL-2.1-or-later
+
 ====
 Mesh
 ====
 
-The mesh subsection of DEM simulations is almost identical to the `CFD <https://chaos-polymtl.github.io/lethe/documentation/parameters/cfd/mesh.html>`_ in Lethe. There are two additional parameters mainly used in DEM and CFD-DEM simulations. These parameters are ``check diamond cells`` and ``expand particle-wall contact search``.
+DEM and CFD simulations read their triangulation through the same routine, so the ``mesh`` subsection of a DEM simulation is the `CFD one <https://chaos-polymtl.github.io/lethe/documentation/parameters/cfd/mesh.html>`_. Every mesh type (``gmsh``, ``dealii`` and the ``lethe`` built-in grids), the mesh transformations (``scale``, ``initial translation``, ``initial rotation axis`` and ``initial rotation angle``), the initial refinement and the refinement in the vicinity of boundaries are all available in ``lethe-particles``.
```
Comment (antonellasuccar) [inline doc/source/parameters/dem/mesh.rst]:
```suggestion
DEM and CFD simulations read their triangulation through the same routine, so the ``mesh`` subsection of a DEM simulation is the same as the `CFD one <https://chaos-polymtl.github.io/lethe/documentation/parameters/cfd/mesh.html>`_. The mesh type (``gmsh``, ``dealii`` and the ``lethe`` built-in grids), the mesh transformations (``scale``, ``initial translation``, ``initial rotation axis`` and ``initial rotation angle``), the initial refinement and the refinement in the vicinity of boundaries are all available in ``lethe-particles``.
```

File: include/core/grids.h [hunk 7]
```diff
@@ -148,6 +152,25 @@ refine_triangulation_at_boundaries(
     }
 }
 
+/**
+ * @brief Build the triangulation which delimits a mesh refinement box
+ *
+ * The box is described by a full set of mesh parameters, so any mesh type
+ * supported by attach_grid_to_triangulation() can be used to delimit a
+ * refinement region. The initial refinement of the box mesh is applied here, so
+ * the resulting triangulation is ready to be used to flag cells.
```
Comment (antonellasuccar) [inline include/core/grids.h:161]: I'm not sure why we need to flag cells if the refinement is done initially

File: include/core/parameters.h (hunk 4 above)
Comment (antonellasuccar) [inline include/core/parameters.h]:
```suggestion
   * @param[in] file_name Name of the parameter file from which the size is
```

File: include/dem/periodic_boundaries_manipulator.h [hunk 8]
```diff
@@ -194,11 +195,12 @@ class PeriodicBoundariesManipulator
   bool periodic_boundaries_enabled;
 
   /**
-   * @brief Direction of the periodic boundaries, it is the perpendicular axis
-   * of the periodic boundaries. Keys of this map are the principal periodic
-   * boundary ids (periodic id 0).
+   * @brief Periodic boundary pairs of the simulation. Each entry carries the
+   * neighbor boundary id of the pair and the direction of periodicity, which
+   * is the perpendicular axis of the periodic boundaries. Keys of this map are
+   * the principal periodic boundary ids (periodic id 0).
```
Comment (antonellasuccar) [inline include/dem/periodic_boundaries_manipulator.h:201]:
```suggestion
   * @brief Periodic boundary pairs of the simulation. Each entry carries the
   * neighbor boundary id of the pair and the direction of periodicity, which
   * is normal to the periodic boundaries. Keys of this map are
   * the principal periodic boundary ids (periodic id 0).
```

File: applications/lethe-parameter-check/lethe_parameter_check.cc [hunk 9]
```diff
@@ -54,18 +54,24 @@ main(int argc, char *argv[])
         }
       else if (solver_family == "lethe-particles")
         {
+          // The DEM solvers do not require any variable size subsection. Their
+          // boundary conditions are declared in the "DEM boundary conditions"
+          // subsection, and a file that declares no manifold leads to
+          // subsections of size zero.
```
Comment (antonellasuccar) [inline applications/lethe-parameter-check/lethe_parameter_check.cc]: Are there more than one DEM solver?
```suggestion
          // The DEM solver does not require any variable size subsection. Its
          // boundary conditions are declared in the "DEM boundary conditions"
          // subsection, and a file that declares no manifold leads to
          // subsections of size zero.
```

File: release_notes/current/2026_08_17_Blais.md
Comment (antonellasuccar) [inline release_notes/current/2026_08_17_Blais.md:1]: Nicely explained!

File: source/core/grids.cc [hunk 10]
```diff
@@ -14,52 +13,87 @@
...
 
 
+/**
+ * @brief Warn when a mesh scaling factor is applied to a grid which carries curved manifolds
+ *
+ * Meshes generated by deal.II (type = dealii) or by Lethe's built-in generators
+ * (type = lethe) may carry curved manifolds attached to the triangulation.
+ * GridTools::scale moves the vertices but does not scale these attached
+ * manifolds, so a scale factor other than 1 can leave the manifolds
+ * inconsistent with the mesh and distort curved geometries under refinement.
```
Comment (antonellasuccar) [inline source/core/grids.cc]:
```suggestion
 * GridTools::scale moves the vertices but does not scale their attached
 * manifolds, so a scale factor other than 1 can leave the manifolds
 * inconsistent with the mesh and distort curved geometries under refinement.
```

File: source/core/grids.cc [hunk 11]
```diff
@@ -14,52 +13,87 @@
...
+ * The scaling is still applied, but the user is warned so that the resulting
+ * mesh can be checked; such geometries can alternatively be dimensioned through
+ * the grid arguments.
```
Comment (antonellasuccar) [inline source/core/grids.cc]:
```suggestion
 * mesh can be checked; such geometries can alternatively be scaled through
 * the grid arguments.
```

File: source/fem-dem/cfd_dem_coupling.cc [hunk 12]
```diff
@@ -1617,7 +1617,8 @@ CFDDEMSolver<dim, PropertiesIndex>::solve()
     this->cfd_dem_simulation_parameters.cfd_parameters.mesh,
     this->cfd_dem_simulation_parameters.cfd_parameters.manifolds_parameters,
     true,
-    this->cfd_dem_simulation_parameters.cfd_parameters.boundary_conditions);
+    this->cfd_dem_simulation_parameters.cfd_parameters.boundary_conditions
+      .periodic_boundaries);
```
Comment (antonellasuccar) [inline source/fem-dem/cfd_dem_coupling.cc:1621]: I don't think that the ``restart`` bool should be hard coded to true. This prevents the meshes in CFD-DEM from getting refined. We can fix this in this PR or I can do it in a subsequent one I am doing to fix the cell-to-size ratio output after a restart.

File: source/fem-dem/cfd_dem_coupling_matrix_free.cc [hunk 13]
```diff
@@ -1637,7 +1637,8 @@ CFDDEMMatrixFree<dim, PropertiesIndex>::solve()
     this->cfd_dem_simulation_parameters.cfd_parameters.mesh,
     this->cfd_dem_simulation_parameters.cfd_parameters.manifolds_parameters,
     true,
-    this->cfd_dem_simulation_parameters.cfd_parameters.boundary_conditions);
+    this->cfd_dem_simulation_parameters.cfd_parameters.boundary_conditions
+      .periodic_boundaries);
```
Comment (antonellasuccar) [inline source/fem-dem/cfd_dem_coupling_matrix_free.cc:1641]: Same thing here

File: tests/dem/combined_periodic_offsets.cc [hunk 14]
```diff
@@ -59,14 +59,17 @@ test()
...
-  // Configure the manipulator with one periodic pair per direction. The
-  // direction map is keyed by the principal periodic boundary id (2*d).
-  std::map<types::boundary_id, unsigned int> periodic_directions;
+  // Configure the manipulator with one periodic pair per direction. The map is
+  // keyed by the principal periodic boundary id (2*d) and matches the pairs
+  // collected above.
+  Parameters::PeriodicBoundaries periodic_boundaries;
   for (int d = 0; d < dim; ++d)
-    periodic_directions[2 * d] = d;
+    periodic_boundaries[2 * d] = {.neighbor_id =
```
Comment (karlwilson) [inline tests/dem/combined_periodic_offsets.cc:67]: Does the indexation `2*d` imply that there can only be 2 boundaries per dimension in the mesh?

**Review by karlwilson — COMMENTED** [review]
Great addition. Reviewed all files, found nothing problematic

File: tests/core/size_of_subsections_01.cc [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+                       "subsection DEM boundary conditions\n"
+                       "  set number of boundary conditions = 2\n"
+                       "end\n");
+
+  const Parameters::SizeOfSubsections dem_sizes =
+    Parameters::get_size_of_subsections(dem_file_name, false);
+
+  deallog << "DEM parameter file: " << dem_sizes.boundary_conditions
+          << " boundary conditions, " << dem_sizes.manifolds << " manifolds"
+          << std::endl;
```
Comment (antonellasuccar) [inline tests/core/size_of_subsections_01.cc:79]: We are printing the max number, so I'm not sure if we need to show boundary conditions and manifolds separately

**Review by antonellasuccar — COMMENTED** [review]
Nice PR! I think at the moment the initial refinement does not work for the CFD-DEM simulations

File: tests/dem/combined_periodic_offsets.cc (hunk 14 above)
Comment (blaisb) [reply tests/dem/combined_periodic_offsets.cc:67]: No, it's just for the purpose of this unit tests that we do this. Not in general in the code :)

**Comment by blaisb** [comment]
> Nice PR! I think at the moment the initial refinement does not work for the CFD-DEM simulations

Nope it does not. There is still an issue open for that, but I will need to make a follow-up PR to fix it.

File: source/fem-dem/cfd_dem_coupling.cc (hunk 12 above)
Comment (blaisb) [reply source/fem-dem/cfd_dem_coupling.cc:1621]: I would do it in a subsequent PR. I wanted this PR not to change any tests or anything.
When we change this, we will change test output so I'd rather do a follow-up.

File: include/core/grids.h (hunk 7 above)
Comment (blaisb) [reply include/core/grids.h:161]: Well we use a "box" to flag the cells which are going to be refined.
The box refinement is a local refinement in space.

File: source/fem-dem/cfd_dem_coupling_matrix_free.cc (hunk 13 above)
Comment (blaisb) [reply source/fem-dem/cfd_dem_coupling_matrix_free.cc:1641]: Yup. I agree, but we will fix it in a follow-up PR. Otherwise tests results and etc. will change.

File: tests/core/size_of_subsections_01.cc (hunk 15 above)
Comment (blaisb) [reply tests/core/size_of_subsections_01.cc:79]: This is just a unit tests, so why not? I'm not sure what do we lose by not printing both?
Especially once we will evolve the DEM boundary conditions to work like the CFD ones, this will be necessary.

**Review by antonellasuccar — APPROVED** [review]

**Review by karlwilson — APPROVED** [review]
